fdbserver/TagPartitionedLogSystem.actor.cpp

```diff
@@ -1279,22 +1331,104 @@ struct TagPartitionedLogSystem : ILogSystem, ReferenceCounted<TagPartitionedLogS
 	}
 
-	virtual Tag getRandomTxsTag() const {
+	Tag getRandomTxsTag() const override {
 		return Tag(tagLocalityTxs, deterministicRandom()->randomInt(0, txsTags));
 	}
 
-	virtual TLogVersion getTLogVersion() const {
+	TLogVersion getTLogVersion() const override {
 		return tLogs[0]->tLogVersion;
 	}
 
+	int getLogRouterTags() const override { return logRouterTags; }
+
+	Version getStartVersion() const override {
+		ASSERT(tLogs.size() > 0);
+		return tLogs[0]->startVersion;
+	}
+
+	std::map<LogEpoch, ILogSystem::EpochTagsVersionsInfo> getOldEpochTagsVersionsInfo() const override {
+		std::map<LogEpoch, EpochTagsVersionsInfo> epochInfos;
+		for (const auto& old : oldLogData) {
+			epochInfos.insert(
+			    { old.epoch, ILogSystem::EpochTagsVersionsInfo(old.logRouterTags, old.epochBegin, old.epochEnd) });
+			TraceEvent("OldEpochTagsVersions", dbgid)
+			    .detail("Epoch", old.epoch)
+			    .detail("Tags", old.logRouterTags)
+			    .detail("BeginVersion", old.epochBegin)
+			    .detail("EndVersion", old.epochEnd);
+		}
+		return epochInfos;
+	}
+
+	inline Reference<LogSet> getEpochLogSet(LogEpoch epoch) const {
+		for (const auto& old : oldLogData) {
+			if (epoch == old.epoch) return old.tLogs[0];
+		}
+		return Reference<LogSet>(nullptr);
+	}
+
+	void setBackupWorkers(const std::vector<InitializeBackupReply>& replies) override {
+		ASSERT(tLogs.size() > 0);
+
+		Reference<LogSet> logset = tLogs[0];  // Master recruits this epoch's worker first.
+		LogEpoch logsetEpoch = this->epoch;
+		oldestBackupEpoch = this->epoch;
+		for (const auto& reply : replies) {
+			Reference<AsyncVar<OptionalInterface<BackupInterface>>> worker(new AsyncVar<OptionalInterface<BackupInterface>>(OptionalInterface<BackupInterface>(reply.interf)));
+			if (reply.backupEpoch != logsetEpoch) {
+				// find the logset from oldLogData
+				logsetEpoch = reply.backupEpoch;
+				oldestBackupEpoch = std::min(oldestBackupEpoch, logsetEpoch);
+				logset = getEpochLogSet(logsetEpoch);
+				ASSERT(logset.isValid());
+			}
+			logset->backupWorkers.push_back(worker);
+		}
+		backupWorkerChanged.trigger();
+	}
+
+	bool removeBackupWorker(const BackupWorkerDoneRequest& req) override {
+		bool removed = false;
+		Reference<LogSet> logset = getEpochLogSet(req.backupEpoch);
+		std::string msg("BackupWorkerNotFound");
+		if (logset.isValid()) {
+			for (auto it = logset->backupWorkers.begin(); it != logset->backupWorkers.end(); it++) {
+				if (it->getPtr()->get().interf().id() == req.workerUID) {
+					msg = "BackupWorkerRemoved";
+					logset->backupWorkers.erase(it);
+					removed = true;
+					break;
+				}
+			}
+		}
+
+		if (removed) {
+			oldestBackupEpoch = epoch;
+			for (const auto& old : oldLogData) {
+				if (old.epoch < oldestBackupEpoch && old.tLogs[0]->backupWorkers.size() > 0) {
+					oldestBackupEpoch = old.epoch;
+				}
+			}
+			backupWorkerChanged.trigger();
+		}
+
+		TraceEvent(msg.c_str(), dbgid)
```

Comment:
make a static name and then log the removed bool

Reply:
Done.